### DISCUSSION TITLE
getting-started/csharp

### DISCUSSION THREAD

**Discussion by giscus** [velopack/velopack.docs#8] [General]
# getting-started/csharp

Get started with .NET 5+ (cross-platform) or .Net Framework.

https://docs.velopack.io/getting-started/csharp

**Comment by liu-super** [comment]
For this line, what should I replace there, all releases or a single file?
```
var mgr = new UpdateManager("https://the.place/you-host/updates");
```

**Reply by caesay** [reply]
The URL (`https://the.place/you-host/updates`) should point to the place you uploaded all of the files produced by `vpk pack`. For example, `https://the.place/you-host/updates/releases.win.json` should be a valid URL.

**Reply by liu-super** [reply]
Wow~ You responded so quickly！thank you very much!

**Comment by SebastianDevelops** [comment]
I get the below error in my CLI even after being prompted that vpk have successfully installed. I'm running windows 11.

```
vpk : The term 'vpk' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the
spelling of the name, or if a path was included, verify that the path is correct and try again.
At line:1 char:1
+ vpk
+ ~~~
    + CategoryInfo          : ObjectNotFound: (vpk:String) [], CommandNotFoundException
    + FullyQualifiedErrorId : CommandNotFoundException
```

**Reply by SebastianDevelops** [reply]
Sorry, I got the answer in the FAQ:
https://docs.velopack.io/troubleshooting/faq

Thanks.

**Comment by bearyung** [comment]
Is there any way for the app to simulate the result of UpdateManager?
e.g. I want to get the result in debug mode of my IDE so I can setup the breakpoint and test my UI logic, but the app will throw the below error:
```
System.Exception: Cannot perform this operation in an application which is not installed.
   at Velopack.UpdateManager.EnsureInstalled() in ./Velopack/UpdateManager.cs:line 442
```

**Reply by caesay** [reply]
We are looking to improve this with an msbuild integration in the near future, but for now, the only way to test UpdateManager is to provide a mocked `IVelopackLocator` to the constructor - there you can mock properties like whether the app is installed, where it's installed to, where Update.exe is, etc.

**Reply by bearyung** [reply]
Thanks for pointing me to the right direction. mocked ```IVelopackLocator``` definitely serve what I need.

**Reply by systemzee** [reply]
Could you dive into this more? I am attempting to create a IVelopackLocator class and its not working out very well, I need to test this in debug.  Sorry very new to C#, only 8 weeks in now so still on the learning curve. I have Velopack working once I publish it, just need to get it in a test/debug state as I implement features for it.  BTW wonderful library!

**Reply by caesay** [reply]
Can you check `TestVelopackLocator` class and see if that solves your problem? If it doesn't, please feel free to join our Discord and I am happy to help 1 on 1.

**Reply by systemzee** [reply]
This did the trick, I completely missed seeing this in the docs! Thanks again! you have a donation link? More than willing to help good developers.

**Reply by caesay** [reply]
There's a donation link on our github repo!

**Comment by DrEsteban** [comment]
Just to confirm, is the idea that one would **manually** call the `UpdateMyApp()` method somewhere in the codebase? (After the call to `VelopackApp.Build().Run()`?) Or does the Velopack framework use Reflection to automatically call this method? I'm just trying to determine how much of this is "automagic" vs providing an example of how one could accomplish auto-updating.

If it's the latter, are there any other Reflection-based hooks one could specify? In either case, it might be good to clarify this in the docs! Along with recommendations on how/when to invoke the auto update method. (Before or after the `Run()`?)

**Reply by Keboo** [reply]
It is the former - nothing is automatic/reflection based. You would place the code from `UpdateMyApp()` somewhere in your application. That code provides the steps needed to perform the update check, download the update, and then restart the application all at once. However, typically a better UX will provide the user with some level of interaction around these steps. These may include, showing when an update is available, allowing the user to choose when to download the update, and allowing the user to choose when to restart the application.

**Reply by DrEsteban** [reply]
Thank you!

**Comment by MindaugasSakalys** [comment]
Hi, does Velopack work with net framework 4.7.2?

**Reply by caesay** [reply]
Yep, anything later or equal to .Net framework 4.6.2 is supported. Also anything dotnet >= 5.0.

**Comment by nabeelbaghoor** [comment]
Hey I hope you will help, I need to avoid installiong the app or putting into the programs, or anywhere. I have make an exe of my code which we put into windows autostart, and it keeps running in background with just a process running in task bar. I just need to implement a way so that it can keep updating itself whenever we push to master of github. No need to install it, nothing.

**Reply by caesay** [reply]
Velopack produces a portable package (`YourAppId-Portable.zip`) which does not need to be installed and is self-updating. Feel free to extract your app from that zip and place it wherever you'd like.

**Reply by nabeelbaghoor** [reply]
Hey thanks for quick response. I extracted the .zip, which i exe i need to use as normal, the one inside current folder or the one oputside that. also, there is an Update exe. i don't need to run it right.

**Reply by nabeelbaghoor** [reply]
for example, if i run the the first exe inside zip (not the one in current folder), it will keep updating itselfg while it will be running whenever i will ask him to do so from code? and to push a new update what i need to do for example from s3. i need to put this zipped folder extracted there? and it my exe will update automatically whenevr it will check s3.

**Reply by caesay** [reply]
You can run the exe in the root, or the one inside the current folder. Either will do. The one at the root of the portable package is just there to act like a shortcut for the user so it is easier to run your app. You do not need to run `Update.exe` manually. Please refer to the "Integrating", "Packaging", and "Distributing" documentation for more info on how to update/distribute your app.

**Comment by RadJKW** [comment]
Using the portable version of the packaged application, ` mgr.ApplyUpdatesAndRestart(newVersion);` has not been able to successfully restarted my application. The application will update and close, requiring me to manually open the app again. 

I have the following folder structure where i've placed the extracted contents of the portable zip from the deployment server 
```
📁 C:\
└─📁 MyAppFolder
   │ 📄 MachineConfig.json
   ├ 📁 Offline_Data
   └ 📁 App
      │    📄 MyApp.exe
      │    📄 Update.exe
      │    📄 Velopack.log
      │    📄 .portable
      ├─ 📁 current
      │    📄 dlls
      │    📄 MyApp.exe
      │    📄 other dotnet files
      └─ 📁 packages
         │  📄 MyApp-0.0.5-full.nupkg
         └─ 📁 VelopackTemp
```

**Reply by caesay** [reply]
Please make sure you are using the same version of `Velopack` and `vpk`. If these two are not the same version you may run into this issue.

**Comment by argenisr0117** [comment]
I have a .NET Framework Windows Form project (target .NET Framework 4.7.2) and when I tried to publish, I got the following error  "error MSB4803: The task "GenerateBootstrapper" is not supported on the .NET Core version of MSBuild. Please use the .NET Framework version of MSBuild. See https://aka.ms/msbuild/MSB4803 for further details. [C:\Projects\CondoSys\CondoSys.Forms\CondoSys.Forms.csproj]" Any suggestions ?

**Reply by caesay** [reply]
Compile your app using whatever is most ergonomic for you. If you are using legacy/outdated csproj formats, you may not be able to use dotnet.exe to compile your app. Feel free to use msbuild.exe, Visual Studio, or whatever else works or you have been using in the past. Or, even better, update your csproj to the latest SDK-style project format. Once you've compiled your app, just run `vpk pack ... -p {path-to-your-build-output} -f net472` to build your Velopack release.

**Reply by argenisr0117** [reply]
Hey @caesay thanks for your response. Now i'm facing another issue. I have a server and I created a site on IIS (https://mysite.com/). Basically I just created a dummy index page and copied the files that vpack created. When I ran vpk download http --url https://mysite.com/  I got the following:
 Downloading release file 'releases.win.json' from 'https://mysite.com/releases.win.json?arch=x64&os=win&rid=win-x64'.
[19:54:41 INF] Found 11 release(s) in remote file
[19:54:41 INF] Downloading CondoSys.Forms-3.0.0-full.nupkg...
[19:54:41 INF] Downloading 'CondoSys.Forms-3.0.0-full.nupkg' from 'https://mysite.com/CondoSys.Forms-3.0.0-full.nupkg'.
[19:54:42 ERR] Response status code does not indicate success: 404 (Not Found)., retrying in 1 second.

It's like the files does not exists but they are in the sites folder..
![image](https://github.com/user-attachments/assets/aacc742a-14d1-4a8a-a838-f3b3d9cd4f26)

What am I missing?

Thanks

**Reply by caesay** [reply]
This is an issue with your IIS configuration and is not related to Velopack, IIS only returns certain known file types by default. If you are struggling with IIS I'd suggest you look into using Azure Storage or Amazon S3 to host your releases, and you can use the `vpk upload az` or `vpk upload s3` command to manage that for you.

**Reply by argenisr0117** [reply]
Hey @caesay I figured out that the .nupkg MIME TYPE was missing in the configuration of my IIS site. After adding it, it worked. Placing my fix here if somebody has the same issue.

Just go to your ISS site and add this entry in the MIME TYPE confg 
File Name extension: .nupkg 
MIME TYPE: application/zip.

Another question..
My app is downloading the update but when it close to install the new version, I am getting a "denied access" like the app has no permission to install itself.

Is there any configuration in velopak to grant permissions or I just have to run the app as administrator ?

**Reply by caesay** [reply]
The documentation comments is not the right place for individual support / questions about your specific integration. The doc comments are mostly for questions which clarify what Velopack is capable of eg. information missing from the docs. For tailored help / ongoing discussion please join our Discord.

**Comment by enisn** [comment]
I need to add Authorization header to all requests to `releases.win.json` and `.nupkg` files. How can I do it in a single point? Does it use HttpClientFactory internally?

**Reply by Keboo** [reply]
That depends on which source you are using. If you are using a simple Http source, there are two injections points `IFileDownloader` and `IUpdateSource`. 

The `UpdateManager` has a constructor that takes in an `IUpdateSource` as a constructor parameter. If you wanted to control all of the calls for retrieving the manifest and the files this interface is what you would implement for complete control. If, however, you just want to leverage the exiting stuff and only modify some of the HTTP calls (which is sounds like is your use case), you can implement the `IFileDownloader` interface and pass it to an instance of `SimpleWebSource`.  The easiest way to do this will be to derive from the existing `HttpClientFileDownloader` and override either the `CreateHttpClientHandler` or `CreateHttpClient`. There is an `authorization` parameter passed to `CreateHttpClient`, however, the current implementation doesn't pass a value down to it.

**Reply by enisn** [reply]
Thanks 🙏 
Everything is ok but checking `relases.*.json` json, I've just created a PR for C# SDK that allows extending methods of `SimpleWebSource`
https://github.com/velopack/velopack/pull/181

**Comment by SebastianDevelops** [comment]
How do I add a space in the name of my application? I tried `My-App-Name`, `My_App_Name` and I tried `My%20App%20Name` but none of these seemed to work.

**Reply by caesay** [reply]
You can use `--packTitle` for your friendly app name.

**Comment by ozdmremrekaan** [comment]
I am getting this error while trying to pack my app even though I have enough permissions
UnauthorizedAccessException: Access to the path 'C:\Users\User02\AppData\Local\Temp\Velopack\temp.1\PreprocessPackDirWin\Prodity.Desktop_ExecutionStub.exe' is denied.

**Reply by ozdmremrekaan** [reply]
Solved it by turning Off firewall

**Reply by caesay** [reply]
Rather than turning off your firewall completely, you could add a rule to exclude scanning in `%temp%/Velopack`

**Reply by ozdmremrekaan** [reply]
Thanks. Now I am having a different problem, When I try to run the exe that is inside of publish folder it does not open but It is working in release doesnt work.

**Comment by wangdasheng360** [comment]
Why isn't WithBeforeUninstallFastCallback is called in my app when uninstalling?

**Reply by caesay** [reply]
The docsite comments are not a place to get support for your application. Check your [Velopack logs]( https://docs.velopack.io/troubleshooting/debugging), and then if you're still having issues please join our Discord for support.

**Comment by cesar-albusac** [comment]
Hello! I successfully tested on a Demo WinForms App. 

However, when I try to add to the production WinForms App, I am not even able to start the app when I add the velopack library and `VelopackApp.Build().Run();`

No logs, only on the Windows Event Viewer I can see these errors:

```
Description: The process was terminated due to an unhandled exception.
Exception Info: System.IO.FileNotFoundException
   at CMS.Ui.Desktop.CMSDesktopLoader.Main(System.String[])

Faulting application name: OfficeEvolveDesktop.exe, version: 5.9.2407.25155, time stamp: 0xc81ac9a9
Faulting module name: KERNELBASE.dll, version: 10.0.19041.4717, time stamp: 0x16b0a901
Exception code: 0xe0434352
Fault offset: 0x000000000003b699
Faulting process ID: 0xfb8
Faulting application start time: 0x01db0a762b6cd51e
Faulting application path: C:\PE\onprem\integration\bin\Debug\OfficeEvolveDesktop.exe
Faulting module path: C:\Windows\System32\KERNELBASE.dll
Report ID: f69f0664-1e79-4251-8431-4c0b3ab2a98d
Faulting package full name: 
Faulting package-relative application ID: 
```

Any ideas? Thanks a million

**Reply by cesar-albusac** [reply]
I checked if the file C:\Windows\System32\KERNELBASE.dll exists and it does

**Reply by caesay** [reply]
Add better exception handling, logging, review that your installed application is not missing any required files.

**Reply by cesar-albusac** [reply]
Thanks, it looks like is something in my production app (.Net version or a 3rd party library)

**Comment by cesar-albusac** [comment]
I am trying to exclude packaging a folder by using the --exclude paramenter but I am unable to do it .

What am I doing wrong? 

`vpk pack -u "TestApp" --packVersion 1.1.4 -p . -e MyApp.exe -o C:\OutputPackages --exclude 'Server/.*' `

When I install the generated program exe, I see there is a Server folder on it.

**Reply by cesar-albusac** [reply]
I want to exlude multiple folders apart from the Server one, and also the default pdb files.

**Reply by cesar-albusac** [reply]
Sorted, ifnoring all these folders: 

`--exclude '(^|/|\\)(Server|Config.Clean|Data|DataSets|cs|de|es|fr|it|ja|ko|pl|pt-BR|ru|tr|runtimes|Transforms|WebView2|win|zh-Hans|zh-Hant)(/|\\).*' `

**Comment by vkyamane** [comment]
For this line:
```
var mgr = new UpdateManager("https://the.place/you-host/updates");

```
When using AWS S3, do I point the object to the .json file that have the releases information? Or is this only available with a public bucket?

**Reply by caesay** [reply]
The URL you pass in should not contain the releases.channel.json file. It should be one level up, so if `https://yoursite.com/updates/releases.channel.json` was a valid URL, then you would pass `https://yoursite.com/updates` to UpdateManager.

**Comment by cgilib** [comment]
Hi, I am using Visual Studio with .net framework 4.7.2. When I tried to install vpk from "Manage Nuget Package", I got the following error message,

Package 'vpk 0.0.942' has a package type 'DotnetTool' that is not supported by project 'MyApp'.

Was it because I don't have .net installed? 

Thanks.

**Reply by caesay** [reply]
You need to run `dotnet tool update -g vpk` to install the `vpk` command line tool, so you will need .net 8 installed for that. The `Velopack` package is what you should install into your project `MyApp`.

**Reply by cgilib** [reply]
Hi, I will install .net 8 and give it a go. Thanks so much.

**Comment by Pureaznangel** [comment]
In case that we need to upload update folder (Release) to a host, is there any way to protect the update (files, especially portable zipped file) from unauthorized direct link downloading (e.g password protected, encryption ...) ?

**Reply by cgilib** [reply]
You can enhance the security of the host by using a client certificate. The application code responsible for checking updates should be customized to consistently send the client certificate to the host for authentication.

**Reply by Keboo** [reply]
This is a bit outside the scope of what the built-in providers give you. Another option depending on how you are hosting the files. You could implement a custom `IUpdateSource` that you pass to your `UpdateManager`. You can use the [existing sources as references](https://github.com/velopack/velopack/tree/develop/src/lib-csharp/Sources). Your new update source would need to perform any needed authentication for accessing and downloading the updates.

**Comment by fan-van** [comment]
“Setup Error
An error has occurred: Failed to start application (Therequested operation requires elevation,(os error 740)) .  ”  I set <requestedExecutionLevel level="requireAdministrator" uiAccess="false" /> in the app.manifest, it prompts this error. How can I avoid this prompt?

**Reply by caesay** [reply]
Velopack does not support applications requiring admin at this time.

**Reply by fan-van** [reply]
Thank you very much! Is it possible to not show the prompt after the installation is complete?

**Reply by caesay** [reply]
You will need to remove the admin requirement from your app.manifest, otherwise velopack will not work correctly.

**Comment by Pureaznangel** [comment]
I found an Extension here: https://github.com/krzysztofautomatyk/VelopackExtension.GoogleDrive/tree/master/VelopackExtension.GoogleDrive has the function to store updated files on google drive. But I read it all the time but don't understand how to use it. Can you give me a simple example code to use it? 

Thank you.

**Reply by caesay** [reply]
This is not an official package. Please contact the author of that library directly.

**Comment by Pureaznangel** [comment]
![image](https://github.com/user-attachments/assets/c89559bc-143e-4393-9149-ddf1c2a93fc2)

Because the folder structure of the software changes a lot when using Velopack, I want to move the folders and files in the red square inside the "Current" folder (my app) in the "Portable Release Folder". Is this possible? And if so, is there any way to do that?

Thank you,

**Reply by caesay** [reply]
No. Velopack apps are portable (meaning, they can be moved around by the user) but they must maintain this structure.

**Comment by Devin-goBILDA** [comment]
To preface I am using this in combination with GitHub to host my releases. Is there a good way with Velopack to check if the release that I would like to update to is a GitHub Pre-release? I would like users to be able to toggle a switch that will let them swap back and fourth between this and a regular GitHub Release. (I know i can use channels, but thought this could simplify my workflow)

**Reply by caesay** [reply]
You can set prerelease to true or false in the GithubSource constructor. If false the source will filter out any prereleases and can act as your "stable" channel.

**Reply by Devin-goBILDA** [reply]
I remembered seeing something about that a while ago and forgot to double check before I asked. Thank you for helping out!

**Comment by cgilib** [comment]
Hi,

The vpk tool relies on update.exe but it's not available by simply rebuilding the solution. May I ask where I can get update.exe and files the VPK depends on? I know I can get them by installing the vpk, however I am just wondering if they can be built from the source code like Clowd.Squirrel. 

Thanks.

**Reply by caesay** [reply]
You can compile it with cargo (eg. `cargo build --features windows`) or you can download it from a [recent gh actions build](https://github.com/velopack/velopack/actions/runs/14765557239).

**Reply by cgilib** [reply]
Thank you. I installed Rust and was able to compile and produce update.exe etc.

Now my application is able to download the patches and apply the patches. Just a question about the patch installation, is there a way to pack it so it can be installed silently without asking if I want to replace the older version? I like the way how Clowd.Squirrel installs the patch, no dialog, no background cmd window and just install. Thanks

Regards,

**Reply by caesay** [reply]
It's not a supported option, as users can do this by mistake and overwrite their install dir. If you have the source code, of course you can make any changes you wish. This comment thread is not the place to discuss modifications, if you have further questions about this please join our Discord or comment on the appropriate github issue.

**Reply by cgilib** [reply]
Thank you, 

The following two lines stops made the application not ask anymore, although the cmd window is still showing in the background but it's much better. By doing this way, we can force the user to run the latest version available.

mgr.WaitExitThenApplyUpdates(null, true, true);   
Environment.Exit(0);

**Reply by cgilib** [reply]
And the cmd window is gone when I run in release mode. Thank you so much.

**Comment by Themenor19** [comment]
I'm wanting to host my updates on a private github repo. What would be the best way to go about that with this project? Or is there even a way to get around that?

**Reply by caesay** [reply]
Create a fine-grained access token which has read-only access to your github releases, then pass that token in as a parameter to `GithubSource`. Note that if you compile this token into your application, it will be retrievable by the users of your application, so this is only really useful if your users are trusted users (such as employees of your company or similar). You may also wish to store releases in a separate repo to your source code, for added separation.

**Comment by Silen1t** [comment]
I still didn't understand this code line
```cs
var mgr = new UpdateManager("https://the.place/you-host/updates");
```
specifically the link like can I put my github repo like this "github.com/USERNAME/REPO-NAME"

**Reply by caesay** [reply]
See [my comment here](https://github.com/velopack/velopack.docs/discussions/8?notification_referrer_id=NT_kwDOABOkf7MxMDAzNTk2NjY3MToxMjg3Mjk1#discussioncomment-11169969) for more info. Also you need to use a [GithubSource](https://docs.velopack.io/reference/cs/Velopack/Sources/GithubSource/constructors) if you want to host your releases on Github.

**Reply by Silen1t** [reply]
that was fast thinks for the help.

**Comment by guilhermegchagas** [comment]
Does this work on **MacCatalyst**? I got it working on Windows, but I haven’t found any information about MacCatalyst support.
I was able to build the package, but when I run `VelopackApp.Build().Run()`, I get the error: `OS platform '' is not supported.`

**Reply by caesay** [reply]
We've not done anything intentional to support MacCatalyst. I don't really know how it works, but it looks like it would only require minor changes to support. Feel free to open a feature request via issues on GitHub!

**Comment by rfaccani** [comment]
Hello!
How can I debug my WinForms application and test how updates work? 

I tried and followed all the steps, but when I run the request to check for updates, an error message appears saying that the application is not installed.

Obviously, it is not installed because I am in Debug mode, but even if I install it, nothing changes.

I am attaching the error message.

```
Velopack.Exceptions.NotInstalledException
  HResult=0x80131500
  Message=This operation cannot be performed in an application that is not installed. Please install the application and try again.
```
Thanks!

**Reply by caesay** [reply]
You can only run UpdateManager in an installed application (that means running `vpk pack`, and then running the subsequently generated installer executable). I recommend testing this way is best for most people.

If you want to test in your IDE, you need to provide a custom `IVelopackLocator` to this constructor to tell Velopack where to find required files and folders: `public UpdateManager(string urlOrPath, UpdateOptions? options = null, IVelopackLocator? locator = null)`.

**Reply by rfaccani** [reply]
Ok Thanks, I'll try

**Comment by SamKr** [comment]
My app is getting rejected from the Microsoft Store because after uninstall, there are files that remain in the `AppData\Local\AppName` folder, namely `.dead` and `velopack.log`. Is there a way to disable this and have the folder completely removed? Or, if not, is there a post-uninstall hook I could use in conjunction with a bat file in a temp folder (which I could prepare with the pre-uninstall hook)?

**Reply by caesay** [reply]
This has been fixed in https://github.com/velopack/velopack/pull/606 and will be released soon.
